src/main/java/com/google/sps/servlets/GroupsServlet.java

```diff
@@ -26,74 +30,76 @@
-      ImmutableList<Map<String, String>> groups = ImmutableList.of((Map) group);
-      
-      // Each group has its own map which points to its info and all maps are passed into the template as a list
-      // This will make it easier when groups are queried from Datastore
-      ImmutableMap<String, ImmutableList<Map<String, String>>> data = ImmutableMap.of(GROUPS_KEY, groups);
-
-      String htmlString = SoyRendererUtils.getOutputString(GROUPS_SOY_FILE, GROUPS_TEMPLATE_NAMESPACE, data);
-
-      response.setContentType(CONTENT_TYPE_HTML);
-      response.getWriter().println(htmlString);
-    } catch (Exception entityError) {
-       ServletUtils.printBadRequestError(response, ENTITY_ERROR_MESSAGE);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    
+    Query groupsQuery = new Query(GROUP_KIND);
+    PreparedQuery preparedGroupsQuery = datastore.prepare(groupsQuery); 
+
+    ImmutableList.Builder<ImmutableMap<String, String>> groupsListBuilder = new ImmutableList.Builder<>();
+    for (Entity group : preparedGroupsQuery.asIterable()) {
+      ImmutableMap<String, String> groupMap = ImmutableMap.of(
+          GROUP_NAME_PROPERTY, (String) group.getProperty(GROUP_NAME_PROPERTY),
+          GROUP_IMAGE_PROPERTY, (String) group.getProperty(GROUP_IMAGE_PROPERTY),
+          GROUP_DESCRIPTION_PROPERTY, (String) group.getProperty(GROUP_DESCRIPTION_PROPERTY));
+      groupsListBuilder.add(groupMap);
     }
+
+    ImmutableList<ImmutableMap<String, String>> groupsList = groupsListBuilder.build();
+
+    // Each group has its own map which points to its info and all maps are passed into the template as a list
+    // This will make it easier when groups are queried from Datastore
+    ImmutableMap<String, ImmutableList<ImmutableMap<String, String>>> groupsData = 
+        ImmutableMap.of(GROUPS_KEY, groupsList);
+
+    String groupsPageHtml = SoyRendererUtils.getOutputString(GROUPS_SOY_FILE, GROUPS_TEMPLATE_NAMESPACE, groupsData);
+
+    response.setContentType(CONTENT_TYPE_HTML);
+    response.getWriter().println(groupsPageHtml);
   }
 
   /**
    * Create a group. Parameters are name, description (optional), and image.
    */
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    //TODO: Unhardcode. Only one hardcoded group will be created and displayed temporarily, but 
-    // will change to enable more groups.
-    String groupId = createGroup(GROUP_NAME, GROUP_IMAGE, GROUP_DESCRIPTION);
+    String name = getParameter(request, GROUP_NAME_PROPERTY);
+    String image = getParameter(request, GROUP_IMAGE_PROPERTY);
 
-    response.setContentType(CONTENT_TYPE_HTML);
+    if (Strings.isNullOrEmpty(name) || Strings.isNullOrEmpty(image)){
+      ServletUtils.printBadRequestError(response, GROUPS_BAD_REQUEST_MESSAGE);
+      return;
+    }
+    
+    String groupId = createGroup(name, getParameter(request, GROUP_DESCRIPTION_PROPERTY), image);
```

Comment:
just wondering, is there a reason you didn't pull out the description at the top like you did the name and image?

Reply:
thought the description wasn't mandatory, is it? it's an easy fix regardless

Reply:
yeah each group has to have a name, image, and description. at least that's how I set it up in the HTML. but like you said, it's an easy fix if you guys want to make it optional